File: sysplant/templates/stubs/sanitizer.nim
# Egg hunter sanitizer: patches egg markers with real syscall opcodes at runtime
var SPT_EGG_PATTERN {.global.}: array[8, byte] = [ ##__EGG_BYTES__## ]

when defined(amd64):
    # syscall (0F 05) + nop (90) + nop (90) + ret (C3) + nop (90) + int3 (CC) + int3 (CC)
    var SPT_EGG_REPLACE {.global.}: array[8, byte] = [0x0F'u8, 0x05, 0x90, 0x90, 0xC3, 0x90, 0xCC, 0xCC]
else:
    # sysenter (0F 34) + nop (90) + nop (90) + ret (C3) + nop (90) + int3 (CC) + int3 (CC)
    var SPT_EGG_REPLACE {.global.}: array[8, byte] = [0x0F'u8, 0x34, 0x90, 0x90, 0xC3, 0x90, 0xCC, 0xCC]

proc SPT_SanitizeSyscalls*(): bool =
    var dwOldProtect: DWORD = 0

    # Get base address of the current module
    let pBase = cast[PVOID](GetModuleHandleA(nil))
    if pBase == nil:
        return false

    # Parse PE headers to find .text section
    let pDos = cast[PIMAGE_DOS_HEADER](pBase)
    let pNt = cast[PIMAGE_NT_HEADERS](cast[uint](pBase) + cast[uint](pDos.e_lfanew))
    let pFirstSection = cast[PIMAGE_SECTION_HEADER](cast[uint](pNt) + cast[uint](sizeof(IMAGE_NT_HEADERS)))

    var pText: PVOID = nil
    var szText: SIZE_T = 0

    for i in 0 ..< pNt.FileHeader.NumberOfSections.int:
        let pSection = cast[PIMAGE_SECTION_HEADER](cast[uint](pFirstSection) + cast[uint](i * sizeof(IMAGE_SECTION_HEADER)))
        let sectionName = cast[ptr uint32](addr pSection.Name[0])
        if sectionName[] == 0x7865742E'u32 or sectionName[] == 0x444F432E'u32:
            pText = cast[PVOID](cast[uint](pBase) + cast[uint](pSection.VirtualAddress))
            szText = cast[SIZE_T](pSection.Misc.VirtualSize)
            break

    if pText == nil or szText == 0:
        return false

    let dwEggSize = SPT_EGG_PATTERN.len

    # Make .text section writable
    if VirtualProtect(pText, szText, PAGE_EXECUTE_READWRITE, addr dwOldProtect) == 0:
        return false

    var dwPatched: int = 0

    # Scan for egg pattern and replace
    var i = 0
    while i <= szText.int - dwEggSize:
        var bMatch = true
        for j in 0 ..< dwEggSize:
            if cast[ptr byte](cast[uint](pText) + cast[uint](i + j))[] != SPT_EGG_PATTERN[j]:
                bMatch = false
                break

        if bMatch:
            for j in 0 ..< dwEggSize:
                cast[ptr byte](cast[uint](pText) + cast[uint](i + j))[] = SPT_EGG_REPLACE[j]
            dwPatched += 1
            i += dwEggSize
        else:
            i += 1

    # Restore original protection
    discard VirtualProtect(pText, szText, dwOldProtect, addr dwOldProtect)

    return dwPatched > 0

# Auto-patch egg markers at import time
discard SPT_SanitizeSyscalls()

File: sysplant/templates/stubs/sanitizer.rs
// Egg hunter sanitizer: patches egg markers with real syscall opcodes at runtime

// Sanitizer-specific imports
use windows_sys::Win32::System::Diagnostics::Debug::IMAGE_SECTION_HEADER;
use windows_sys::Win32::System::Memory::{VirtualProtect, PAGE_EXECUTE_READWRITE};

unsafe extern "system" {
    fn GetModuleHandleA(lpModuleName: *const u8) -> isize;
}

static SPT_EGG_PATTERN: [u8; 8] = [ ##__EGG_BYTES__## ];

#[cfg(target_arch = "x86_64")]
// syscall (0F 05) + nop (90) + nop (90) + ret (C3) + nop (90) + int3 (CC) + int3 (CC)
static SPT_EGG_REPLACE: [u8; 8] = [0x0f, 0x05, 0x90, 0x90, 0xC3, 0x90, 0xCC, 0xCC];

#[cfg(target_arch = "x86")]
// sysenter (0F 34) + nop (90) + nop (90) + ret (C3) + nop (90) + int3 (CC) + int3 (CC)
static SPT_EGG_REPLACE: [u8; 8] = [0x0f, 0x34, 0x90, 0x90, 0xC3, 0x90, 0xCC, 0xCC];

pub unsafe fn spt_sanitize_syscalls() -> bool {
    let mut dw_old_protect: u32 = 0;

    // Get base address of the current module
    let h_module = GetModuleHandleA(core::ptr::null());
    if h_module == 0 {
        return false;
    }
    let p_base = h_module as *const u8;

    // Parse PE headers to find .text section
    let p_dos = p_base as *const IMAGE_DOS_HEADER;
    let p_nt = p_base.offset((*p_dos).e_lfanew as isize) as *const IMAGE_NT_HEADERS64;
    let p_first_section = (p_nt as *const u8).offset(core::mem::size_of::<IMAGE_NT_HEADERS64>() as isize) as *const IMAGE_SECTION_HEADER;

    let mut p_text: *const u8 = core::ptr::null();
    let mut sz_text: usize = 0;

    for i in 0..(*p_nt).FileHeader.NumberOfSections as isize {
        let p_section = p_first_section.offset(i);
        let section_name = *((*p_section).Name.as_ptr() as *const u32);
        if section_name == 0x7865742E || section_name == 0x444F432E {
            p_text = p_base.offset((*p_section).VirtualAddress as isize);
            sz_text = (*p_section).Misc.VirtualSize as usize;
            break;
        }
    }

    if p_text.is_null() || sz_text == 0 {
        return false;
    }

    let egg_size = SPT_EGG_PATTERN.len();

    // Make .text section writable
    if VirtualProtect(
        p_text as *const _,
        sz_text,
        PAGE_EXECUTE_READWRITE,
        &mut dw_old_protect,
    ) == 0 {
        return false;
    }

    let mut patched: u32 = 0;

    // Scan for egg pattern and replace
    let mut i: usize = 0;
    while i <= sz_text - egg_size {
        let mut found = true;
        for j in 0..egg_size {
            if *p_text.offset((i + j) as isize) != SPT_EGG_PATTERN[j] {
                found = false;
                break;
            }
        }

        if found {
            let p_write = p_text as *mut u8;
            for j in 0..egg_size {
                *p_write.offset((i + j) as isize) = SPT_EGG_REPLACE[j];
            }
            patched += 1;
            i += egg_size;
        } else {
            i += 1;
        }
    }

    // Restore original protection
    VirtualProtect(
        p_text as *const _,
        sz_text,
        dw_old_protect,
        &mut dw_old_protect,
    );

    patched > 0
}
